Add Solana Foundational Modules

enoldev opened the pull request.

enoldev committed:
WIP

enoldev committed:
Add doc

enoldev committed:
Add doc

enoldev committed:
Remove clones

enoldev committed:
Remove clone

enoldev committed:
Update README

maoueh commented on solana-common/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+version = "0.1.0"
+description = "Substreams showcasing the building blocks of Solana"
+edition = "2021"
+repository = "https://github.com/streamingfast/substreams-explorers"
```
Wrong repo.

maoueh commented on solana-common/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+
+[dependencies]
+bs58 = "0.4"
+hex-literal = "0.3"
```
Don't think we need this

maoueh commented on solana-common/proto/instructions.proto:
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
```
@enoldev I was discussing with Charles and Ed and they think we should start with only transaction. They think that getting instructions alone is not as useful because usually you also need to access other instruction within the same transaction which is not easily modelled now.

Let's discuss this Monday in greater details.

maoueh commented on solana-common/src/instruction.rs:
```diff
@@ -0,0 +1,67 @@
+        let acct_keys: Vec<&[u8]> = confirmed_transaction.resolved_accounts().iter().map(|k| k.as_slice()).collect();
+
+        if let Some(transaction) = &confirmed_transaction.transaction {
+            for instruction in confirmed_transaction.all_instructions().into_iter() {
```
```suggestion
            for instruction in confirmed_transaction.all_instructions().iter() {
```

Should be good.

enoldev replied:
It seems the method is not available, only `into_iter()`

maoueh replied:
Try `for instruction in confirmed_transaction.all_instructions() {` it should work

maoueh commented on solana-common/proto/instructions.proto:
```diff
@@ -0,0 +1,14 @@
+message Instruction {
+  string program_id = 1;
+  repeated string accounts = 2;
+  string data = 3;
```
If we keep instruction I think this one should be `bytes` directly, data is byte centric and it will make consumption for the IDL much easier after since it will read the byte directly and avoid the extra string to bytes decoding.

maoueh commented on solana-common/src/instruction.rs:
```diff
@@ -0,0 +1,67 @@
+    let filtered_instructions = instructions.instructions
+            .into_iter()
+            .filter(|inst| {
+                let mut keys = Vec::new();
+                keys.push(format!("program:{}", inst.program_id));
+
+                matches_keys_in_parsed_expr(&keys, &query).expect("matching events from query")
```
```suggestion
                let keys = vec![format!("program:{}", inst.program_id)];
                matches_keys_in_parsed_expr(&, &query).expect("matching events from query")
```

maoueh commented on solana-common/src/pb/sf.solana.type.v1.rs:
```diff
@@ -0,0 +1,256 @@
+// @generated
```
I think that now that we use the one from `substreams-solana`, we don't need that.

enoldev commented on solana-common/substreams.yaml:
```diff
@@ -0,0 +1,103 @@
+      type: proto:sf.solana.type.v1.Block
+    doc: |
+      `map_block_without_votes` allows you to consume a full Solana Block without Vote instructions (`Vote111111111111111111111111111111111111111`). 
+      If you consume it on HISTORICAL data (+1000 block from HEAD), you will be reading the StreamingFast cache, thus savings costs on the amount of TB read.
```
```suggestion
      If you consume it on HISTORICAL data (+1000 blocks from HEAD), you will be reading from the StreamingFast cache, thus saving costs on the amount of TB read.
```

enoldev commented on solana-common/src/transaction.rs:
```diff
@@ -0,0 +1,41 @@
+            }).map(|inst| (inst.tx_hash, true))
+            .collect::<HashMap<String, bool>>();
+
+    let filtered_transations: Vec<ConfirmedTransaction> = transactions.transactions
```
```suggestion
    let filtered_transactions: Vec<ConfirmedTransaction> = transactions.transactions
```

enoldev committed:
Fixes

enoldev committed:
Remove filtered instructions module

maoueh commented on solana-common/src/block.rs:
```diff
@@ -0,0 +1,31 @@
+use substreams_solana::pb::sf::solana::r#type::v1::Block;
+
+// Vote111111111111111111111111111111111111111
+static VOTE_INSTRUCTION: &'static [u8] = &[7, 97, 72, 29, 53, 116, 116, 187, 124, 77, 118, 36, 235, 211, 189, 179, 216, 53, 94, 115, 209, 16, 67, 252, 13, 163, 83, 128, 0, 0, 0, 0];
+
```
We have `bs58!` now:

```suggestion
static VOTE_INSTRUCTION: [u8; 32] = b58!("Vote111111111111111111111111111111111111111");
```

You will need to change the `contains` condition below too to be `message.account_keys.iter().any(|v| v == &VOTE_INSTRUCTION)`

maoueh committed:
Make code leaner and reformatted a few files, updated to latest `substreams-solana`

maoueh committed:
Refactored using latest unreleased `substreams-solana` and `substreams-rs`

- Using `ExprMatcher` to avoid parsing the query for each match
- Updated to use latest `substreams-solana`
- Usage of `Address` type where it made sense.

sduchesneau committed:
solana: rework logic to exclude redundant mappers
